module/state_synchronization/blob_io.go

```diff
@@ -0,0 +1,270 @@
+package state_synchronization
+
+import (
+	"errors"
+	"io"
+	"sync"
+
+	"github.com/onflow/flow-go/network"
+)
+
+var ErrClosedBlobChannel = errors.New("send/receive on closed blob channel")
+
+type blobChannel struct {
+	blobs chan network.Blob
+	once  sync.Once
+	done  chan struct{}
+}
+
+func (bc *blobChannel) Send(blob network.Blob) error {
+	select {
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case bc.blobs <- blob:
+		return nil
+	case <-bc.done:
+		return ErrClosedBlobChannel
+	}
+}
+
+func (bc *blobChannel) Receive() (network.Blob, error) {
+	select {
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	default:
+	}
+
+	select {
+	case blob := <-bc.blobs:
+		return blob, nil
+	case <-bc.done:
+		return nil, ErrClosedBlobChannel
+	}
+}
+
+func (bc *blobChannel) Close() error {
+	bc.once.Do(func() { close(bc.done) })
```

Comment:
I don't think it's necessary, and it could cause a panic in Send / Receive if something is already blocking on read.

For context, my implementation is based on Go's `Pipe`: https://cs.opensource.google/go/go/+/refs/tags/go1.17.3:src/io/pipe.go;l=39